### ISSUE TITLE
how to set threshold values for modelConfidenceThreshold ,modelScoreThreshold , modelNMSThreshold for yolov8 c++ for custom trained model

### ISSUE THREAD

**Issue by ranjith502** [ultralytics/ultralytics#5737]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

i have trained yolov8 model on detecting smoking( cigarette ) . when i inference on images using python it is performing well . it is drawing one boundary box for one image only .  i converted the trained model into onnx format and doing inference on images using c++ code provided by the analytics. below is the c++ code of yolov8 i used 

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-CPP-Inference

when i use c++ code iam getting 3 boundary boxes , i think the problem is with the setting of  modelConfidenceThreshold ,modelScoreThreshold , modelNMSThreshold in c++ code. i have used default values only . how to set the correct values so that it will work like python version.

![Screenshot from 2023-10-16 17-13-34](https://github.com/ultralytics/ultralytics/assets/85454586/894c328c-0b13-4e4a-9a67-86c2a271e2df)
![Screenshot from 2023-10-16 17-06-34](https://github.com/ultralytics/ultralytics/assets/85454586/88cb49b8-7d98-4171-b020-04c6fbf1229c)

below are the default values for my c++ code , 

    float modelConfidenceThreshold {0.25};
    float modelScoreThreshold      {0.45};
    float modelNMSThreshold        {0.50};

for reference i have attacched , args.yml file ,precision and recall curves. i have played with some values for nms ,confidence, by reference form args.yml file but no use.
![val_batch2_pred](https://github.com/ultralytics/ultralytics/assets/85454586/e61fd056-7a2d-450c-ba43-bd35d18c164e)
![results](https://github.com/ultralytics/ultralytics/assets/85454586/18bc20ba-0d55-4b39-aff1-85c65f1dbe74)
![R_curve](https://github.com/ultralytics/ultralytics/assets/85454586/5a49a2bc-60a5-48b1-84a7-1b0a04697df3)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/85454586/779970e4-f9b8-491e-8717-6e35239111d3)
![P_curve](https://github.com/ultralytics/ultralytics/assets/85454586/f176f597-2a28-4caa-826e-d126e5fecc48)
![F1_curve](https://github.com/ultralytics/ultralytics/assets/85454586/452cdafb-916d-4162-946e-710019200c7f)
 
 below is my args.yml file
![Screenshot from 2023-10-16 17-31-28](https://github.com/ultralytics/ultralytics/assets/85454586/68f7eb35-8a9a-440f-a4fc-71268d6e70dd)
![Screenshot from 2023-10-16 17-31-18](https://github.com/ultralytics/ultralytics/assets/85454586/9bf73263-1ea9-45d9-aeba-58303aec1646)

**Comment by glenn-jocher** [comment]
@ranjith502 hello, 

Thank you for providing such detailed information; this makes it easier to understand your issue. The difference between the outputs of the Python version and C++ version of YOLOv8 you're observing could indeed be due to the different threshold values.

The `modelConfidenceThreshold`, `modelScoreThreshold` and `modelNMSThreshold` parameters adjust the sensitivity of your detection model. 

- `modelConfidenceThreshold` is the minimum probability to filter weak detections. A lower confidence threshold will result in detecting more objects but increases the chance of false positives. 

- `modelScoreThreshold` is used to filter out detections that have a combined confidence score less than the threshold value. As with the confidence threshold, a lower score threshold will result in more detections but increases the chance of false positives. 

- `modelNMSThreshold` is used in non-maximum suppression to provide a threshold below which overlapping bounding boxes will be removed. A bounding box will be removed in non-maximum suppression if it has an intersection over union (IoU) value greater than this threshold with another bounding box that has a higher confidence score. 

You can fine-tune these values based on the precision and recall curves from your model's evaluation to balance the trade-off between precision (fewer false positives) and recall (capturing more true positives). Start with small increments and check the performance at each step to identify the optimal setting that helps to make the results closer to your Python version. 

Please note that due to the nature of how C++ and Python handle float numbers, slight differences might still exist between the predictions of the same model using the two different programming languages.

I hope this helps and encourages you to continue experimenting!

Best,
Glenn

**Comment by ranjith502** [comment]
@glenn-jocher , thanks for the detailed ans sir ,

may i know how to take modelConfidenceThreshold, modelScoreThreshold and modelNMSThreshold values form precision and recall curves ,any reference link for blogs or videos

**Comment by glenn-jocher** [comment]
Hello @ranjith502,

I'm glad that you found the previous explanation helpful.

Regarding your first question, precision and recall curves typically visualize the trade-off between two metrics: precision and recall. Higher precision means fewer false positives, and higher recall means fewer false negatives. 

Typically, the confidence threshold is varied to generate these curves. If you have a higher requirement for precision (fewer false positives), you might want to increase your thresholds. On the other hand, if recall is more of a priority (fewer missed true positives), you might choose to lower the thresholds. 

However, it's important to note that these thresholds aren't necessarily derived directly from these curves. Instead, they're given values based on the requirements of your specific use-case and determined empirically - what works best given the data and the task at hand. 

As for your second query, when you're running inference using your model in Python, these threshold values are likely set as arguments to the detection function in your YOLOv8 code. These values control the sensitivity of your model during inference. 

Adjusting these values in your C++ inference code to match the Python-version thresholds should provide similar results. However, due to potential differences in language handling and floating-point computations, the results might not be identical but should be reasonably close.

In summary, using the Python-version threshold values can be a good starting point, but you might need to adjust them further based on performance and requirements of your specific use case, which could be informed by your precision-recall curves.

I hope this helps clarify your questions a bit further!

Best,
Glenn

**Comment by Farhanbhai023** [comment]
hello @glenn-jocher 
I am also working same project but Maine use case is different i have also trained the model with one class only previously i trained the model with 3 classes which is present in default classes.txt file in the yolov8 GitHub repository but know i trained the in yolov8 python it's give me good  results of object detection in python but when i am trying to convert the model from Py torch to onnx for inferencing with c++ it's give me the results with three bounding boxes in output image i don't know how to give the customized  modelConfidenceThreshold, modelScoreThreshold and modelNMSThreshold parameters for my model give some reference links of article or videos that i can follow and resolve my error and explain me how to give the customized parameters for my model "modelConfidenceThreshold, modelScoreThreshold and modelNMSThreshold".

**Comment by glenn-jocher** [comment]
@Farhanbhai023 dear user,

I understand you're encountering an issue when converting your PyTorch model to ONNX for inference in C++. You're observing three bounding boxes in the output image instead of just one. This could be related to the `modelConfidenceThreshold`, `modelScoreThreshold` and `modelNMSThreshold` parameters.

These three parameters essentially control the model's sensitivity during the inference. 

- `modelConfidenceThreshold` is a minimum probability to filter out detection results with low reliability. Lower values might lead to more detections but risk having false positives.

- `modelScoreThreshold` filters out detections with a combined confidence score less than the stated value. Just like the confidence threshold, too low a value here might make the model more sensitive but at the cost of false positives.

- `modelNMSThreshold` provides a threshold in non-maximum suppression, over which overlapping bounding boxes will be removed.

The way to adjust these parameters is usually more empirical rather than analytical, as it largely depends on your specific use-case. If you want fewer false positives, but potentially miss some true positives, you can increase your thresholds. If you want to minimize missed true positives at the risk of more false positives, then you decrease your thresholds.

In your C++ code, these threshold values should be set to closely match what you've used with your Python detection function. This can result in a more similar performance between your Python and C++ inferences. Do note however that slight variations may not be completely avoidable due to differences in language implementation.

I hope this was useful. Please do not hesitate to ask if you have further questions.

Best,
Maria
